cabot/metricsapp/api/elastic.py

```diff
@@ -1,11 +1,54 @@
 from elasticsearch import Elasticsearch
+from enum import Enum
+
+
+class SupportedMetrics(Enum):
+    # Response includes the field "value"
+    METRICS_SINGLE = ['min', 'max', 'avg', 'value_count', 'sum', 'cardinality', 'moving_avg', 'derivative']
+    # Response includes the field "values"
+    METRICS_MULTIPLE = ['percentiles']
 
 
 def create_es_client(urls, timeout):
     """
     Create an elasticsearch-py client
     :param urls: comma-separated string of urls
+    :param timeout: timeout for queries to the client
     :return: a new elasticsearch-py client
     """
     urls = [url.strip() for url in urls.split(',')]
     return Elasticsearch(urls, timeout=timeout)
+
+
+def validate_query(query):
```

Comment:
Would this be a simpler / more readable function?

```python
def validate_query(query, msg_prefix='Elasticsearch query format error'):
    """
    Validate that an Elasticsearch query is in the format we want
    (all aggregations named 'agg', 'date_histogram' most internal
    aggregation, other metrics named the same thing as their metric
    type, e.g. max, min, avg...).
    :param query: the raw Elasticsearch query
    """

    # Loop through all the aggregations, stopping when we hit a date_histogram
    query = query.get('aggs', None)

    if query is None:
        return

    query = query.get('agg', None)

    if query is None:
        raise ValueError('{}: boo'.format(msg_prefix))

    if 'date_histogram' not in query:
        validate_query(query)
        return

    # Last case. We cannot go any deeper than this
    if 'agg' in query:
        raise ValueError('{}: date_histogram must be the innermost'
                         'aggregation (besides metrics)'.format(msg_prefix))

    # If we found a date_histogram the rest of the aggs should be metrics
    query = query.get('aggs', None)

    if query is None:
        raise ValueError('{}: query must include a metric'.format(msg_prefix))

    for metric, item in query.iteritems():
        if metric not in METRICS_SINGLE + METRICS_MULTIPLE:
            raise ValueError('{}: unsupported metric {}'.format(
                msg_prefix, metric))

        if metric not in item:
            raise ValueError('{}: metric name must be the same '
                             'as the metric type'.format(msg_prefix))
```

Issues I saw with the prefix function:
* We repeat the key fetch operations
```python
if not query.get('agg'):
   ...

query = query['agg']
```
We are doing the key get twice in two or three places. Not good for performance.
* using `iteritems()` is more efficient than iterating through keys and doing a get.